modules/heat_conduction/src/actions/RadiationTransferAction.C

```diff
@@ -0,0 +1,336 @@
+                 "provided too");
+
+    std::vector<boundary_id_type> fixed_T_boundary_ids =
+        getParam<std::vector<boundary_id_type>>("fixed_temperature_sidesets");
+
+    std::vector<FunctionName> fixed_T_funcs =
+        getParam<std::vector<FunctionName>>("fixed_boundary_temperatures");
+
+    // check length of fixed_boundary_temperatures
+    if (fixed_T_funcs.size() != fixed_T_boundary_ids.size())
+      mooseError("Size of parameter fixed_boundary_temperatures and fixed_temperature_sidesets "
+                 "must be equal.");
+
+    std::vector<BoundaryName> fixed_T_boundary_names;
+    std::vector<FunctionName> fixed_T_function_names;
+    for (unsigned int k = 0; k < fixed_T_boundary_ids.size(); ++k)
+    {
+      boundary_id_type bid = fixed_T_boundary_ids[k];
+
+      // find the right entry in _boundary_ids
+      auto it = std::find(_boundary_ids.begin(), _boundary_ids.end(), bid);
+
+      // check if entry was found: it must be found or an error would occur later
+      if (it == _boundary_ids.end())
+        mooseError("Fixed temperature sideset ", bid, " not present in sidesets.");
+
+      // this is the position in the _boundary_ids vector; this is what
+      // we are really after
+      auto index = std::distance(_boundary_ids.begin(), it);
+
+      // collect the correct boundary names
+      for (auto & e : radiation_patch_names[index])
+      {
+        fixed_T_boundary_names.push_back(e);
+        fixed_T_function_names.push_back(fixed_T_funcs[k]);
+      }
+    }
+    params.set<std::vector<BoundaryName>>("fixed_temperature_boundary") = fixed_T_boundary_names;
+    params.set<std::vector<FunctionName>>("fixed_boundary_temperatures") = fixed_T_function_names;
+  }
+
+  // the view factor userobject name
+  params.set<UserObjectName>("view_factor_object_name") = viewFactorObjectName();
+
+  // this userobject needs to be executed on linear and timestep end
+  ExecFlagEnum exec_enum = MooseUtils::getDefaultExecFlagEnum();
+  exec_enum = {EXEC_LINEAR, EXEC_TIMESTEP_END};
+  params.set<ExecFlagEnum>("execute_on") = exec_enum;
+
+  // add the object
+  _problem->addUserObject("ViewFactorObjectSurfaceRadiation", radiationObjectName(), params);
+}
+
+std::vector<std::vector<std::string>>
+RadiationTransferAction::radiationPatchNames() const
+{
+  std::vector<std::vector<std::string>> radiation_patch_names(_boundary_ids.size());
+  for (unsigned int j = 0; j < _boundary_ids.size(); ++j)
+  {
+    boundary_id_type bid = _boundary_ids[j];
+    std::string base_name = _mesh->getBoundaryName(bid);
+    std::vector<std::string> bnames;
+    for (unsigned int i = 0; i < _n_patches[j]; ++i)
+    {
+      std::stringstream ss;
+      ss << base_name << "_" << i;
+      bnames.push_back(ss.str());
+    }
+    radiation_patch_names[j] = bnames;
+  }
+  return radiation_patch_names;
+}
+
+void
+RadiationTransferAction::addMeshGenerator() const
+{
+  MultiMooseEnum partitioners = getParam<MultiMooseEnum>("partitioners");
```

Comment:
I don't like that there's a second action where you can add mesh generators. This is also the source of the problem with the heat conduction breakage, but I'd have to look into why it's happening.

It appears that you would like the final mesh generator name so that you don't need to supply the input to your PatchSidesetGenerator. However, as you've discovered, this is tricky since you have to execute the mesh_generator tree to get the last mesh generator name, so you've had to duplicate a lot of logic from different Actions to try to get everything to work. It all feels very hackish. I would suggest that as a start you just force the user to supply the final generator name in your Action. Then all of this code just goes away and so does your error in the HeatConduction app. I'm willing to look at this a bit more with you to see if we can come up with a reasonable way to eliminate the need to supply the final name in the input file.

Reply:
> Providing the last mesh generator name is a reasonable solution.

Wait, are you saying that you are willing to do that for now? If so, that means you can drop the "add_mesh_generator_complete" task and the associated change in the HeatConduction module. I suspect that this will just fix the odd failure you are seeing and will also cleanup the tasks a bit so we don't have two tasks that allow the addition of mesh_generators.

I'll think about this problem a bit and see if we can come up with another solution to determining the final generator programmatically at the right time so you won't require that input in the future.